test/integration/server-selection/server_selection.test.ts

```diff
@@ -0,0 +1,248 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { Collection, MongoClient } from '../../../src';
+import { ConnectionPool } from '../../../src/cmap/connection_pool';
+
+const TEST_METADATA: MongoDBMetadataUI = {
+  requires: {
+    topology: 'single',
+    mongodb: '>=4.0.0'
+  }
+};
+
+describe('Server Selection', function () {
+  let client: MongoClient;
+  let collection: Collection;
+
+  beforeEach(async function () {
+    client = await this.configuration.newClient().connect();
+    collection = client.db('server-selection-operation-count').collection('collection0');
+    await collection.insertMany([{ name: 'joe' }, { name: 'smith' }]);
+  });
+
+  afterEach(async function () {
+    sinon.restore();
+    await collection.deleteMany({});
+    await client.close();
+    client = undefined;
+  });
+
+  context('operationCount', function () {
+    context('operationCount is adjusted properly on successful operation', function () {
+      it('is zero after a successful command', TEST_METADATA, async function () {
+        const server = Array.from(client.topology.s.servers.values())[0];
+        expect(server.s.operationCount).to.equal(0);
+        const commandSpy = sinon.spy(server, 'command');
+
+        await collection.insertOne({
+          name: 'Joe'
+        });
+
+        expect(commandSpy.called).to.be.true;
+        expect(server.s.operationCount).to.equal(0);
+      });
+
+      it('is zero after a successful getMore', TEST_METADATA, async function () {
+        const cursor = collection.find({}, { batchSize: 1 });
+        await cursor.next(); // initialize the cursor
+
+        const server = Array.from(client.topology.s.servers.values())[0];
+        expect(server.s.operationCount).to.equal(0);
+
+        const getMoreSpy = sinon.spy(server, 'getMore');
+
+        await cursor.next();
+
+        expect(getMoreSpy.called).to.be.true;
+        expect(server.s.operationCount).to.equal(0);
+
+        await cursor.close();
+      });
+
+      it('is zero after a successful killCursors', TEST_METADATA, async function () {
+        const cursor = collection.find({}, { batchSize: 1 });
+        await cursor.next(); // initialize the cursor
+
+        const server = Array.from(client.topology.s.servers.values())[0];
+        expect(server.s.operationCount).to.equal(0);
+
+        const killCursorsSpy = sinon.spy(server, 'killCursors');
+
+        await cursor.close();
+
+        expect(killCursorsSpy.called).to.be.true;
+        expect(server.s.operationCount).to.equal(0);
+      });
+    });
+
+    context('operationCount is adjusted properly when operations fail', function () {
+      afterEach(async function () {
+        await client.db('admin').command({
```

Comment:
Kind of a preference but can you pull out the failpoints to something named for the intention, like this one is to disable. And use the FailPoint type annotation so future readers can hover for more info

Reply:
Unless you feel strongly about this, I'm going to leave them inline.  I don't think that pulling them into variables increases readability substantially.  I will type them with the `FailPoint` interface though.

Reply:
Eating my own words and making the change lol